### ISSUE TITLE
Provide isless

### ISSUE THREAD

**Issue by haampie** [JuliaArrays/StaticArrays.jl#367]
For tuples there is

```julia
(1, 2) < (1, 3) #true
```

Would it make sense to have it for SVectors as well?

```julia
SVector(1, 2) < SVector(1, 3) # ERROR: MethodError: no method matching isless(::SVector{2,Int64}, ::SVector{2,Int64})
```

My use case is to sort a long Vector of small SVector elements.

**Comment by mschauer** [comment]
I prefer not, the lexical order is not a natural order for vectors.

**Comment by haampie** [comment]
Do you have an example in mind?

**Comment by mschauer** [comment]
In ℝⁿ the expression x ≤ y almost exclusively refers to the partial order (I have a bias here though).

**Comment by andyferris** [comment]
We strictly follow `Base` in these matters - doing so aids in code compatibility between static and dynamically sized arrays in the long run.

Your best chance would be to open an issue on the main Julia webpage. While lexical ordering is not a natural operation on vector spaces, you might argue it makes just as much sense on a “list” as a “string” (in some languages they are the same (strings being just a list of characters), and `Vector` is our list type).

**Comment by haampie** [comment]
Actually we have lexicographical `isless` for vectors in Julia 0.7 now, so reopening this!

**Comment by andyferris** [comment]
Yep - thanks for checking that. As an implementation detail you might consider overloading `cmp`.

I note that the existing `Base` definition could be made more efficient. If we compare the length of the vectors before the loop (find the minimum), use indices instead of `zip`, and, use `@inbounds`, we probably get something closer to what we could squeeze out of fully unrolled code via a generated function anyway (either way involves branching).

**Comment by mschauer** [comment]
I was also wrong on how it should be: `isless` tests "whether x is less than y, according to a canonical total order" and the lexical order could well be considered the canonical *total* order on vectors.

**Comment by andyferris** [comment]
Yes, that’s right. `isless` doesn’t necessarily provide semantic meaning but rather some way of implementing sorted search and so-on. It could be a Hilbert curve or some-such and it would still be ok, but lexical order works for a variety of element types, is fast, and matches `String` (which in many ways has some similarity with `Vector{Char}`).

**Comment by c42f** [comment]
Well, this is no longer a bug (base implements a fallback), but the generated code for `SVector` is fairly horrible compared to Tuple.

Would be a good first issue for someone to implement this by just calling `Tuple` on the input SVectors and passing that to the comparison.
